File: dsl-api/src/main/java/cbs/dsl/api/EventDefinition.java
package cbs.dsl.api;

import cbs.dsl.api.EventTypes.EventInput;
import cbs.dsl.api.EventTypes.EventOutput;

import java.util.Collections;

/**
 * Defines an event — a triggered operation that orchestrates transactions, context, and finish
 * behaviour.
 *
 * <p>Implementations are typically created via the Kotlin DSL {@code event { }} block or annotated
 * with {@link DslComponent} for compile-time registration.
 */
public interface EventDefinition extends StandardDslDefinition {

  /**
   * Executes this event with the given input.
   *
   * <p>The default implementation returns an empty output. Generated Definition wrappers delegate
   * to the underlying {@link EventFunction}.
   *
   * @param input the event input
   * @return the event output
   */
  default EventOutput execute(EventInput input) {
    return EventOutput.success(Collections.emptyMap());
  }

  default EventOutput preview(EventInput input) {
    return execute(input);
  }
}

File: dsl-api/src/main/kotlin/cbs/dsl/api/EventDefinition.kt
package cbs.dsl.api

import cbs.dsl.api.context.DisplayScope
import cbs.dsl.api.context.EnrichmentContext
import cbs.dsl.api.context.FinishContext
import cbs.dsl.api.context.TransactionsScope

interface EventDefinition {
  val code: String
  val parameters: List<ParameterDefinition>
  val contextBlock: (EnrichmentContext) -> Unit
  val displayBlock: (DisplayScope) -> Unit
  val transactionsBlock: (suspend TransactionsScope.() -> Unit)?
  val finishBlock: (FinishContext, Throwable?) -> Unit
}
